### DISCUSSION TITLE
With nullable value types like `int?`, add automatic unwrapping if the valaue isn't `null`

### DISCUSSION THREAD

**Discussion by WillPittenger** [dotnet/csharplang#7690] [Language Ideas]
Currently, with a type like `string?`, you can use the following to test for null:
```CS
string? value = something;
if(value != null)
  value; // Do something with value as it isn't null and is safe to access
```
But with a value type like `int?`, you need a different syntax:
```CS
int? value = something;
if(value.HasValue)
  value.Value; // The value is non-null and value.Value holds the value.
```
It'd be much better if nullable value types could use the first syntax.  The compiler would then map that to the second if needed.

**Comment by HaloFour** [comment]
You can already use `if (value != null)` to test an nullable value type.

As for the `Value` property, that's necessary as nullable value types are a different type from the actual value type, e.g. `Nullable<int>` vs. `int`.  Silently treating the nullable value type as the value type would create ambiguities, particularly where the value type it wrapped also had a property called `Value`.

**Reply by WillPittenger** [reply]
In the case of a conflicting property, the compiler would attempt to use the one from the value type, in this case `int`.

**Reply by HaloFour** [reply]
That's what makes it ambiguous.  This has been suggested before but the language team was not interested in introducing such an ambiguity.  I do not have a link to the other discussion off-hand at the moment.

**Reply by HaloFour** [reply]
Here's some discussions on the subject:

https://github.com/dotnet/csharplang/issues/1865
https://github.com/dotnet/csharplang/discussions/4647

**Reply by TahirAhmadov** [reply]
I think if we were doing everything from scratch, it could potentially be possible to drop the `.Value` dereferencing. But at this point, given the legacy of existing code, it's downright impossible.

**Comment by lcsondes** [comment]
Instead of making `value` turn into something else inside the body of the `if`, my proposal #9790 is much more limited in scope: it would only allow known-safe implicit conversions from `T?` to `T`. In this case, `value` would firmly remain `int?`, but since it's known to not be null, it would be assignable to `int`, passed to `int` parameters, etc.

These currently don't even build, so no existing code would be affected.

The `T.Value` ambiguity would not be an issue, either. There is no implicit conversion involved to access `.Value`, so it would naturally retain the existing behavior.

**Reply by jnm2** [reply]
Would this silently change meaning? (Probably indicating a well-behavedness issue on the part of the program if the change was problematic?)
```cs
int? v = 1;
if (v != null) M(v);

void M(int? v) => Console.WriteLine("reached before language change");
void M(int v) => Console.WriteLine("reached after language change");
```

**Reply by jnm2** [reply]
Also, would this code continue to compile, or would the non-lifted `+` operator be chosen, inferring `IEnumerable<int>` as the return type of the Select call?
```cs
IEnumerable<int?> M(int? v)
{
    if (v is null) return [];

    return Enumerable.Range(1, 10).Select(n => n + v);
}
```

**Reply by lcsondes** [reply]
I don't think it would change meaning. Not having to do an implicit conversion means the nullable overload is a better candidate, same as this:

```cs
[Flags] enum MyFlags;

Foo(0); // this will pick the int overload

void Foo(MyFlags f) => Console.WriteLine("enum");
void Foo(int i) => Console.WriteLine("int");
void Foo(double d) => Console.WriteLine("double");
```

Lifted operators already include an implicit T? to T conversion to reach the original operator, so they would be unaffected.

**Reply by jnm2** [reply]
Would this change meaning?
```cs
int? v = 1;
if (v != null) M(v);

void M(long? v) => Console.WriteLine("A");
void M(int v) => Console.WriteLine("B");
```

**Reply by lcsondes** [reply]
If the language designers feel that doing so is the better choice. I would not mind if my proposed conversion was last in line.

**Comment by colejohnson66** [comment]
Currently, this already compiles and prints `int?`. Your request would be a breaking change.

```cs
int? x = ...;
if (x is int) // or `x != null`
   Foo(x);

void Foo(int? x) => Console.WriteLine("int?");
void Foo(int x) => Console.WriteLine("int");
```

**Reply by timcassell** [reply]
Why would that break? Exact type matches are always chosen over implicit casts iirc.

**Reply by colejohnson66** [reply]
It definitely could be written that way, and @lcsondes' request would handle it. But OP explicitly requested the opposite, if I'm reading this right:

https://github.com/dotnet/csharplang/discussions/7690#discussioncomment-7576390

> In the case of a conflicting property, the compiler would attempt to use the one from the value type, in this case `int`.

**Reply by franchyd** [reply]
If the possible breaking change was palatable to the language team I think it would be better to make it breaking change and switch those calls.   Conceptually int and int? (while different types at runtime) are just "int that can't be null"" and "int can be null"   And if a function with otherwise identical signature exists, either the programmer is designing bad API, or it would indeed be preferable to use the one binding to "int that can't be null" because the compiler knows in fact that it is not null.  I think the number of cases where this actually causes problems would be low.

**Reply by lcsondes** [reply]
I'm in a weird position, because my proposal #9790 got marked as a duplicate of this one. I agree, it's mostly a duplicate, but I wanted to propose something far more limited in scope than this one, since it was going nowhere.

We'll have to see how confusing things will get if people start talking about multiple different approaches under one ticket.
